[BUG-14514] - Switch frame to an iframe in a Shadow DOM

vishnuv688 opened the pull request:
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Polish (an improvement for an existing feature)
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (improvements to the project's docs)
- [ ] Specification changes (updates to WebDriver command specifications)
- [ ] Internal updates (everything related to internal scripts, governance documentation and CI files)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added the necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Backport Request

[//]: # (The current `main` branch is the development branch for WebdriverIO v9. If your change should be released to the current major version of WebdriverIO (v8), please raise another PR with the same changes against the `v8` branch.)

- [ ] This change is solely for `v9` and doesn't need to be back-ported
- [ ] Back-ported PR at `#XXXXX`

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

christian-bromann commented on packages/webdriverio/src/utils/shadowDom.ts:
```diff
@@ -0,0 +1,27 @@
+    urlFragment: string
+): Promise<WebdriverIO.Element | null> {
+    return await browser.execute(function (fragment: string) {
+        function findIframe(root: ShadowRoot | Document): HTMLIFrameElement | null {
+            const allElements = Array.from(root.querySelectorAll('iframe'))
+            for (const el of allElements) {
+                if (el instanceof HTMLIFrameElement && el.src.includes(fragment)) {
+                    return el
+                }
+            }
+
+            const shadowHosts = Array.from(root.querySelectorAll('*'))
+            for (const host of shadowHosts) {
+                if ((host as HTMLElement).shadowRoot) {
+                    const result = findIframe((host as HTMLElement).shadowRoot!)
+                    if (result) {return result}
+                }
+            }
+
+            return null
+        }
+
+        return findIframe(document)
```
I suggest to move this section into `packages/webdriverio/src/scripts` and remove this file instead.

vishnuv688 replied:
Done

christian-bromann reviewed (commented):
One suggestion

christian-bromann commented on packages/webdriverio/src/scripts/shadowDom.ts:
```diff
@@ -0,0 +1,27 @@
+    browser: WebdriverIO.Browser,
+    urlFragment: string
+): Promise<WebdriverIO.Element | null> {
+    return await browser.execute(function (fragment: string) {
```
scripts should only contain what is being executed in browser, e.g.

```ts
export default function findIframeInShadowDOM (fragment: string) {
  ...
}
```

and then used like this:

```ts
import { findIframeInShadowDOM } from '../../scripts/shadowDom.js'

await browser.execute(findIframeInShadowDOM, urlFragment)
```

vishnuv688 commented:
> Can you please provide some information about the actual bug or problem you are trying to solve here?

Thanks for the review! Just to clarify, this PR is actually fixing two separate bugs we ran into when working with switchFrame() in BiDi mode, especially in cases where the iframe is inside a Shadow DOM.

1. elementId gets lost when using shadow$
The first issue was with waitForExist() — when an element is found via a custom selector strategy like shadow$, it gets re-queried internally using this.parent.$(this.selector). The problem is that this re-query doesn't re-enter the shadow DOM, so the resolved element loses its elementId. That causes a BiDi error later during switchToFrame().

To fix this, I added a check so that we only re-resolve the element if elementId isn’t already set. That way, if we already have a valid element (from shadow$ or similar), we don’t overwrite it.

2. BiDi can’t find iframes inside Shadow DOM
The second issue is that browsingContextLocateNodes() doesn’t support searching inside shadow roots. So if your iframe lives inside a shadow DOM, BiDi just won’t find it with a regular locator.

To work around this, I added a fallback that uses browser.execute() to run a script (findIframeInShadowDOM) which manually traverses the DOM and any shadow roots to find the iframe based on its src.

christian-bromann commented:
> If you'd still prefer a separate reproduction case using [guinea-pig](https://github.com/webdriverio/guinea-pig), I’m happy to create one.

Yes, let's do that. This will help prevent regressing on this side in the future.

> Also, for adding an E2E test case in the WebdriverIO monorepo, could you please confirm where exactly I should place it?

We have already some `switchFrame` tests here: https://github.com/webdriverio/webdriverio/blob/main/e2e/wdio/headless/test.e2e.ts#L549

> And recently, I am facing an issue with the 'create-wdio' package `Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'create-wdio' imported from webdriverio/packages/wdio-cli/build/index.js` during pnpm run build or when trying to run a test case locally.

Mhm .. maybe try update your branch. We don't see these issues in `main` and in our pipelines.

vishnuv688 commented:
> Yes, let's do that. This will help prevent regressing on this side in the future.
Can you please provide me with access to the guinea-pig repo to raise a PR?

christian-bromann commented:
> can you please provide me with access to the guinea-pig repo to raise a PR?

The repo is public, you can raise a PR to it.

vishnuv688 commented:
> > can you please provide me with access to the guinea-pig repo to raise a PR?
> 
> The repo is public, you can raise a PR to it.

I tried but I am getting error.

`git push --set-upstream origin bug-14514
ERROR: Permission to webdriverio/guinea-pig.git denied to vishnuv688.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.`

vishnuv688 commented:
> > can you please provide me with access to the guinea-pig repo to raise a PR?
> 
> The repo is public, you can raise a PR to it.

@christian-bromann I tried but I am getting error.

`git push --set-upstream origin bug-14514
ERROR: Permission to webdriverio/guinea-pig.git denied to vishnuv688.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.`

christian-bromann commented:
@vishnuv688 please fork the repository and push your commits to your fork first, then raise a PR from your fork onto the main project.

christian-bromann commented:
@vishnuv688 I've added the @webdriverio/browserstack team to the project, you should be able to push to the branch now

vishnuv688 commented:
> @vishnuv688 I've added the @webdriverio/browserstack team to the project, you should be able to push to the branch now

@christian-bromann, I have merged the e2e test in [guinea-pig](https://github.com/webdriverio/guinea-pig). Please do review the changes.

christian-bromann commented on e2e/wdio/headless/test.e2e.ts:
```diff
@@ -665,6 +665,30 @@ describe('main suite 1', () => {
+
+            it('should switch to iframe inside shadow root via element', async () => {
+                const host = await browser.$('#wrapper')
+                const iframe = await host.shadow$('iframe')
```
Why do we fetch this with `shadow$`? Wouldn't a normal user just do
```suggestion
                const iframe = await host.$('iframe')
```

christian-bromann reviewed (commented):
One question, code overall looks good to me, great work!

christian-bromann commented:
Mind taking a look at the unit tests as well?

vishnuv688 commented:
> Mind taking a look at the unit tests as well?

@christian-bromann Yes, I've already added unit tests for both key scenarios:
- When the element has a valid elementId, switchFrame uses it directly.
- When the element is unresolved (missing elementId), it re-resolves the selector and then switches.

I have already mentioned the reason for this change in the bug report.

christian-bromann commented:
```

⎯⎯⎯⎯⎯⎯ Failed Tests 10 ⎯⎯⎯⎯⎯⎯⎯

 FAIL  packages/webdriverio/tests/commands/browser/action.test.ts > action command > resolves not resolved wdio elements in pointer action
SyntaxError: "undefined" is not valid JSON
 ❯ packages/webdriverio/tests/commands/browser/action.test.ts:213:21
    211|         const calls = vi.mocked(fetch).mock.calls
    212|         const [,,, [, performActionParam]] = calls as any
    213|         expect(JSON.parse(performActionParam.body)).toMatchSnapshot()

       |                     ^
    214|     })
  Snapshots  1 failed
    215| 
             2 obsolete

             ↳ packages/webdriverio/tests/commands/browser/action.test.ts
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/10]⎯
               · action command > resolves not resolved wdio elements in pointer action 1

               · action command > resolves not resolved wdio elements in wheel action 1
 FAIL  packages/webdriverio/tests/commands/browser/action.test.ts > action command > resolves not resolved wdio elements in wheel action

SyntaxError: "undefined" is not valid JSON
 Test Files  3 failed | 291 passed (294)
 ❯ packages/webdriverio/tests/commands/browser/action.test.ts:222:21
      Tests  10 failed | 3421 passed | 7 skipped (3438)
    220|         const calls = vi.mocked(fetch).mock.calls
   Start at  23:38:57
    221|         const [,,, [, performActionParam]] = calls as any
   Duration  111.86s (transform 3.88s, setup 1.70s, collect 36.06s, tests 148.52s, environment 1.54s, prepare 11.30s)
    222|         expect(JSON.parse(performActionParam.body)).toMatchSnapshot()

       |                     ^

    223|     })
Error: SyntaxError: "undefined" is not valid JSON
 ❯ packages/webdriverio/tests/commands/browser/action.test.ts:213:21


    224| })

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/10]⎯

 FAIL  packages/webdriverio/tests/commands/element/execute.test.ts > execute test > should execute the script
TypeError: Cannot read properties of undefined (reading '1')
 ❯ packages/webdriverio/tests/commands/element/execute.test.ts:26:56
     24|         expect((vi.mocked(fetch).mock.calls[1][0] as any).pathname)
     25|             .toBe('/session/foobar-123/element')
     26|         expect(JSON.parse(vi.mocked(fetch).mock.calls[4][1]?.body as s…
       |                                                        ^
     27|             script: expect.stringContaining('return ((elem, a, b, c) =…
     28|             args: [{ [ELEMENT_KEY]: 'some-elem-123', ELEMENT: 'some-el…

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[3/10]⎯

 FAIL  packages/webdriverio/tests/commands/element/execute.test.ts > execute test > should wait for element to exist
Error: Snapshot `execute test > should wait for element to exist 1` mismatched

- Expected
+ Received

  [
    "http://localhost:4321/session",
    "http://localhost:4321/session/foobar-123/element",
    "http://localhost:4321/session/foobar-123/elements",
-   "http://localhost:4321/session/foobar-123/element",
    "http://localhost:4321/session/foobar-123/execute/sync",
  ]

 ❯ packages/webdriverio/tests/commands/element/execute.test.ts:41:66
     39| 
     40|         await browser.$('#foo').execute((elem) => elem)
     41|         expect(vi.mocked(fetch).mock.calls.map(call => call[0])).toMat…
       |                                                                  ^
     42|           [
     43|             "http://localhost:4321/session",

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[4/10]⎯

 FAIL  packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts > selectByVisibleText test > should select value by visible text
TypeError: Cannot read properties of undefined (reading '0')
 ❯ packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts:45:45
     43|         )
     44|         // @ts-expect-error mock implementation
     45|         expect(vi.mocked(fetch).mock.calls[5][0]!.pathname)
       |                                             ^

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[9/10]⎯

 FAIL  packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts > selectByVisibleText test > should convert number to string when selecting
TypeError: Cannot read properties of undefined (reading '0')
 ❯ packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts:145:45
    143|             .toBe(`${optionSelection}|${optgroupSelection}`)
    144|         // @ts-expect-error mock implementation
    145|         expect(vi.mocked(fetch).mock.calls[5][0]!.pathname)
       |                                             ^
    146|             .toBe('/session/foobar-123/element/some-sub-elem-321/click…
    147|         expect(getElementFromResponseSpy).toBeCalledWith({

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[10/10]⎯


Error: SyntaxError: "undefined" is not valid JSON
 ❯ packages/webdriverio/tests/commands/browser/action.test.ts:222:21



Error: TypeError: Cannot read properties of undefined (reading '1')
 ❯ packages/webdriverio/tests/commands/element/execute.test.ts:26:56



Error: Error: Snapshot `execute test > should wait for element to exist 1` mismatched

- Expected
+ Received

  [
    "http://localhost:4321/session",
    "http://localhost:4321/session/foobar-123/element",
    "http://localhost:4321/session/foobar-123/elements",
-   "http://localhost:4321/session/foobar-123/element",
    "http://localhost:4321/session/foobar-123/execute/sync",
  ]

 ❯ packages/webdriverio/tests/commands/element/execute.test.ts:41:66



Error: TypeError: Cannot read properties of undefined (reading '0')
 ❯ packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts:45:45



Error: TypeError: Cannot read properties of undefined (reading '0')
 ❯ packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts:65:45



Error: TypeError: Cannot read properties of undefined (reading '0')
 ❯ packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts:85:45



Error: TypeError: Cannot read properties of undefined (reading '0')
 ❯ packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts:105:45



Error: TypeError: Cannot read properties of undefined (reading '0')
 ❯ packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts:125:45



Error: TypeError: Cannot read properties of undefined (reading '0')
 ❯ packages/webdriverio/tests/commands/element/selectByVisibleText.test.ts:145:45


 ELIFECYCLE  Command failed with exit code 1.
Error: Process completed with exit code 1.
```

christian-bromann reviewed (approved):
LGTM 👍 

Great job!

christian-bromann commented:
Yes, we will put out a release soon.
